File: app/src/main/java/com/android/sunshine/app/model/WeatherContract.java
package com.android.sunshine.app.model;

import android.content.ContentUris;
import android.net.Uri;
import android.provider.BaseColumns;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

public final class WeatherContract {

    public static final String CONTENT_AUTHORITY = "com.android.sunshine.app";
    public static final Uri BASE_CONTENT_URI = Uri.parse("content://" + CONTENT_AUTHORITY);
    public static final String PATH_WEATHER = "weather";
    public static final String DATE_FORMAT = "yyyyMMdd";

    private WeatherContract() { }

    public static String getDbDateString(final Date date) {
        final SimpleDateFormat dateFormat = new SimpleDateFormat(DATE_FORMAT, Locale.getDefault());
        return dateFormat.format(date);
    }

    public static final class WeatherEntry implements BaseColumns {

        public static final Uri CONTENT_URI = BASE_CONTENT_URI.buildUpon().appendPath(PATH_WEATHER).build();
        public static final String CONTENT_TYPE = "vnd.android.cursor.dir/" + CONTENT_AUTHORITY + "/" + PATH_WEATHER;
        public static final String CONTENT_ITEM_TYPE = "vnd.android.cursor.item/" + CONTENT_AUTHORITY + "/" + PATH_WEATHER;

        public static final String TABLE_NAME = "weather";
        public static final String COLUMN_DATE = "date";
        public static final String COLUMN_WEATHER_ID = "weather_id";
        public static final String COLUMN_SHORT_DESC = "short_desc";
        public static final String COLUMN_MIN_TEMP = "min";
        public static final String COLUMN_MAX_TEMP = "max";
        public static final String COLUMN_HUMIDITY = "humidity";
        public static final String COLUMN_PRESSURE = "pressure";
        public static final String COLUMN_WIND_SPEED = "wind";
        public static final String COLUMN_DEGREES = "degrees";
        public static final String COLUMN_LATITUDE = "latitude";
        public static final String COLUMN_LONGITUDE = "longitude";
        public static final String COLUMN_CITY = "city";
        public static final String COLUMN_LOCATION_SETTINGS = "locSettings";

        public static Uri buildWeatherUri(final long id) {
            return ContentUris.withAppendedId(CONTENT_URI, id);
        }

        public static Uri buildWeatherLocation(final String locationSetting) {
            return CONTENT_URI.buildUpon().appendPath(locationSetting).build();
        }

        public static Uri buildWeatherLocationWithStartDate(final String locationSetting, final long startDate) {
            return CONTENT_URI.buildUpon().appendPath(locationSetting).appendQueryParameter(COLUMN_DATE, String.valueOf(startDate)).build();
        }

        public static Uri buildWeatherLocationWithDate(final String locationSetting, final long date) {
            return CONTENT_URI.buildUpon().appendPath(locationSetting).appendPath(String.valueOf(date)).build();
        }

        public static String getLocationSettingFromUri(final Uri uri) {
            return uri.getPathSegments().get(1);
        }

        public static long getDateFromUri(final Uri uri) {
            return Long.parseLong(uri.getPathSegments().get(2));
        }

        public static String getStartDateFromUri(final Uri uri) {
            return uri.getQueryParameter(COLUMN_DATE);
        }
    }
}

File: app/src/main/kotlin/com/android/sunshine/app/model/WeatherContract.kt
package com.android.sunshine.app.model

import android.content.ContentUris
import android.net.Uri
import android.provider.BaseColumns

import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.Date

public class WeatherContract {

    public class WeatherEntry : BaseColumns {
        class object {

            public val CONTENT_URI: Uri = BASE_CONTENT_URI.buildUpon().appendPath(PATH_WEATHER).build()

            public val CONTENT_TYPE: String = "vnd.android.cursor.dir/" + CONTENT_AUTHORITY + "/" + PATH_WEATHER
            public val CONTENT_ITEM_TYPE: String = "vnd.android.cursor.item/" + CONTENT_AUTHORITY + "/" + PATH_WEATHER
            public val TABLE_NAME: String = "weather"

            public val COLUMN_LOC_KEY: String = "location_id"
            public val COLUMN_DATETEXT: String = "date"
            public val COLUMN_WEATHER_ID: String = "weather_id"
            public val COLUMN_SHORT_DESC: String = "short_desc"
            public val COLUMN_MIN_TEMP: String = "min"
            public val COLUMN_MAX_TEMP: String = "max"
            public val COLUMN_HUMIDITY: String = "humidity"
            public val COLUMN_PRESSURE: String = "pressure"
            public val COLUMN_WIND_SPEED: String = "wind"
            public val COLUMN_DEGREES: String = "degrees"

            public fun buildWeatherUri(id: Long): Uri {
                return ContentUris.withAppendedId(CONTENT_URI, id)
            }

            public fun buildWeatherLocation(locationSetting: String): Uri {
                return CONTENT_URI.buildUpon().appendPath(locationSetting).build()
            }

            public fun buildWeatherLocationWithStartDate(locationSetting: String, startDate: String): Uri {
                return CONTENT_URI.buildUpon().appendPath(locationSetting).appendQueryParameter(COLUMN_DATETEXT, startDate).build()
            }

            public fun buildWeatherLocationWithDate(locationSetting: String, date: String): Uri {
                return CONTENT_URI.buildUpon().appendPath(locationSetting).appendPath(date).build()
            }

            public fun getLocationSettingFromUri(uri: Uri): String {
                return uri.getPathSegments().get(1)
            }

            public fun getDateFromUri(uri: Uri): String {
                return uri.getPathSegments().get(2)
            }

            public fun getStartDateFromUri(uri: Uri): String {
                return uri.getQueryParameter(COLUMN_DATETEXT)
            }
        }
    }

    public class LocationEntry : BaseColumns {
        class object {

            public val CONTENT_URI: Uri = BASE_CONTENT_URI.buildUpon().appendPath(PATH_LOCATION).build()
            public val CONTENT_TYPE: String = "vnd.android.cursor.dir/" + CONTENT_AUTHORITY + "/" + PATH_LOCATION
            public val CONTENT_ITEM_TYPE: String = "vnd.android.cursor.item/" + CONTENT_AUTHORITY + "/" + PATH_LOCATION
            public val TABLE_NAME: String = "location"

            public val COLUMN_LOCATION_SETTING: String = "location_setting"
            public val COLUMN_CITY_NAME: String = "city_name"
            public val COLUMN_COORD_LAT: String = "coord_lat"
            public val COLUMN_COORD_LONG: String = "coord_long"
            public fun buildLocationUri(id: Long): Uri {
                return ContentUris.withAppendedId(CONTENT_URI, id)
            }
        }
    }

    class object {

        public val CONTENT_AUTHORITY: String = "com.android.sunshine.app"
        public val BASE_CONTENT_URI: Uri = Uri.parse("content://" + CONTENT_AUTHORITY)
        public val PATH_WEATHER: String = "weather"
        public val PATH_LOCATION: String = "location"
        public val DATE_FORMAT: String = "yyyyMMdd"

        public fun getDbDateString(date: Date): String {
            val dateFormat = SimpleDateFormat(DATE_FORMAT)
            return dateFormat.format(date)
        }

        public fun getDateFromDb(date: String): Date? {
            val simpleDateFormat = SimpleDateFormat(DATE_FORMAT)
            try {
                return simpleDateFormat.parse(date)
            } catch (e: ParseException) {
                e.printStackTrace()
                return null;
            }
        }
    }
}